File: backend/app/schemas/marketplace.py
"""Marketplace related schemas."""
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# --- Marketplace Item ---

class CreateMarketplaceItemRequest(BaseModel):
    asset_type: str = Field(..., description="Asset type: agent/knowledge_base/workflow")
    asset_id: str = Field(..., description="ID of the source asset")
    title: str = Field(..., min_length=1, max_length=200)
    summary: str = Field("", max_length=500)
    description: str = ""
    cover_image: Optional[str] = None
    category: str = ""
    tags: List[str] = []
    visibility: str = Field("tenant", description="private/department/tenant/public")


class UpdateMarketplaceItemRequest(BaseModel):
    title: Optional[str] = None
    summary: Optional[str] = None
    description: Optional[str] = None
    cover_image: Optional[str] = None
    category: Optional[str] = None
    tags: Optional[List[str]] = None
    visibility: Optional[str] = None


class MarketplaceItemResponse(BaseModel):
    id: str
    tenant_id: str
    creator_id: str
    asset_type: str
    asset_id: str
    title: str
    summary: str
    description: str
    cover_image: Optional[str] = None
    category: str
    tags: List[str] = []
    visibility: str
    status: str
    version: int
    avg_rating: float
    rating_count: int
    usage_count: int
    clone_count: int
    featured: bool
    promoted_level: Optional[str] = None
    creator_name: Optional[str] = None
    creator_tenant_name: Optional[str] = None
    published_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class MarketplaceListItemResponse(BaseModel):
    id: str
    asset_type: str
    title: str
    summary: str
    cover_image: Optional[str] = None
    category: str
    tags: List[str] = []
    avg_rating: float
    rating_count: int
    usage_count: int
    clone_count: int
    featured: bool
    promoted_level: Optional[str] = None
    creator_tenant_name: Optional[str] = None
    published_at: Optional[datetime] = None


# --- Review ---

class SubmitForReviewRequest(BaseModel):
    asset_type: str
    asset_id: str
    title: str = Field(..., min_length=1, max_length=200)
    summary: str = Field("", max_length=500)
    description: str = ""
    cover_image: Optional[str] = None
    category: str = ""
    tags: List[str] = []
    visibility: str = "tenant"


class ReviewActionRequest(BaseModel):
    comment: str = ""


class MarketplaceReviewResponse(BaseModel):
    id: str
    item_id: str
    tenant_id: str
    submitter_id: str
    reviewer_id: Optional[str] = None
    review_type: str
    status: str
    comment: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    created_at: datetime
    # Enriched fields
    item_title: Optional[str] = None
    submitter_name: Optional[str] = None


# --- Rating ---

class CreateRatingRequest(BaseModel):
    score: int = Field(..., ge=1, le=5)
    comment: Optional[str] = None


class MarketplaceRatingResponse(BaseModel):
    id: str
    item_id: str
    user_id: str
    score: int
    comment: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    user_name: Optional[str] = None


# --- Clone ---

class CloneResponse(BaseModel):
    id: str
    source_item_id: str
    target_tenant_id: str
    target_asset_id: str
    cloner_id: str
    created_at: datetime


# --- Admin ---

class FreezeRequest(BaseModel):
    reason: str = ""


class TakedownRequest(BaseModel):
    reason: str = ""


class PromoteRequest(BaseModel):
    target_level: str = Field(..., description="Target visibility: department/tenant/group")


class MarketplaceStatsResponse(BaseModel):
    total_items: int
    published_items: int
    pending_review_items: int
    total_ratings: int
    total_clones: int
    total_usage: int
    avg_rating: float
    items_by_category: Dict[str, int]
    items_by_status: Dict[str, int]
    items_by_tenant: Dict[str, int]

File: frontend/src/types/marketplace.ts
export interface MarketplaceItem {
  id: string;
  tenant_id: string;
  creator_id: string;
  asset_type: 'agent' | 'knowledge_base' | 'workflow';
  asset_id: string;
  title: string;
  summary: string;
  description: string;
  cover_image?: string;
  category: string;
  tags: string[];
  visibility: string;
  status: string;
  reject_reason?: string;
  version: number;
  avg_rating: number;
  rating_count: number;
  usage_count: number;
  clone_count: number;
  featured: boolean;
  promoted_level?: string;
  creator_name?: string;
  creator_tenant_name?: string;
  published_at?: string;
  created_at: string;
  updated_at: string;
}

export interface MarketplaceListItem {
  id: string;
  asset_type: string;
  title: string;
  summary: string;
  cover_image?: string;
  category: string;
  tags: string[];
  avg_rating: number;
  rating_count: number;
  usage_count: number;
  clone_count: number;
  featured: boolean;
  promoted_level?: string;
  creator_tenant_name?: string;
  published_at?: string;
}

export interface MarketplaceRating {
  id: string;
  item_id: string;
  user_id: string;
  score: number;
  comment?: string;
  created_at: string;
  updated_at: string;
  user_name?: string;
}

export interface MarketplaceReview {
  id: string;
  item_id: string;
  tenant_id: string;
  submitter_id: string;
  reviewer_id?: string;
  review_type: string;
  status: string;
  comment?: string;
  reviewed_at?: string;
  created_at: string;
  item_title?: string;
  submitter_name?: string;
}

export interface MarketplaceStats {
  total_items: number;
  published_items: number;
  pending_review_items: number;
  total_ratings: number;
  total_clones: number;
  total_usage: number;
  avg_rating: number;
  items_by_category: Record<string, number>;
  items_by_status: Record<string, number>;
  covered_organizations: number;
  items_by_tenant: Record<string, number>;
}

export interface MarketplaceTrends {
  period_days: number;
  daily_items: Array<{ date: string; count: number }>;
  daily_ratings: Array<{ date: string; count: number; avg_score: number }>;
  daily_clones: Array<{ date: string; count: number }>;
}

// Tool Marketplace types
export interface ToolMarketplaceItem {
  id: string;
  name: string;
  display_name: string;
  description: string;
  category: string;
  tags: string[];
  version: string;
  author: string;
  author_avatar?: string;
  icon?: string;
  homepage?: string;
  repository?: string;
  license?: string;
  install_count: number;
  avg_rating: number;
  rating_count: number;
  featured: boolean;
  verified: boolean;
  status: 'active' | 'deprecated' | 'beta';
  config_schema?: Record<string, unknown>;
  examples?: ToolExample[];
  changelog?: string;
  created_at: string;
  updated_at: string;
  installed?: boolean;
}

export interface ToolExample {
  title: string;
  description: string;
  input: Record<string, unknown>;
  expected_output?: string;
}

export interface ToolMarketplaceRating {
  id: string;
  tool_id: string;
  user_id: string;
  user_name?: string;
  score: number;
  comment?: string;
  created_at: string;
}

export const TOOL_CATEGORIES = [
  'AI',
  'Data',
  'Integration',
  'Automation',
  'Communication',
  'Storage',
  'Analytics',
  'Security',
  'Utility',
] as const;

export type ToolCategory = typeof TOOL_CATEGORIES[number];
